### PR TITLE
Copy subtransforms to output of translations.pipeline_from_stages()

### PR DESCRIPTION
**Pull request by yifanmai**
`translations.pipeline_from_stages()` takes in a pipeline proto and a list of stages, and outputs a pipeline proto in which the transforms specified in the stages have been copied form the input pipeline proto. This change recursively finds all subtransforms that these transforms depend on, and copies them from the input pipeline proto to the output pipeline proto.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by yifanmai** [comment]
R: @robertwb

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 1]
```diff
@@ -524,6 +524,22 @@ def add_parent(child, parent):
         add_parent(parent, parents.get(parent))
       components.transforms[parent].subtransforms.append(child)
 
+  copied_subtransform_ids = set()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:527]: Given that the set of subtransformation is a tree (no cycles) we can get rid of this.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 2]
```diff
@@ -541,6 +557,14 @@ def add_parent(child, parent):
     components.transforms[transform_id].CopyFrom(transform)
     add_parent(transform_id, stage.parent)
 
+  if partial:
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:560]: I think we only want to copy sub-transforms of the transforms from line 552 above, perhaps move this logic up there? (Just a thought, maybe having the method be called (`copy_subtransforms` and having it copy (recursively) just the subtransforms could make this even easier to follow).

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py (hunk 1 above)
Comment (yifanmai) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py]: Thanks! This simplifies things.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py (hunk 2 above)
Comment (yifanmai) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py]: Moved it back up and cleaned up the method.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 3]
```diff
@@ -540,6 +549,8 @@ def add_parent(child, parent):
     transform_id = unique_name(components.transforms, stage.name)
     components.transforms[transform_id].CopyFrom(transform)
     add_parent(transform_id, stage.parent)
+    if partial:
+      copy_subtransforms(transform)
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:553]: Nit: You can move this up into the `if partial` block above.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 4]
```diff
@@ -1475,6 +1482,12 @@ def only_element(iterable):
   return element
 
 
+def only_transform(stage):
+  # type: Stage -> beam_runner_api_pb2.PTransform
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py]: 14:14:31 apache_beam/runners/portability/fn_api_runner/translations.py:1485: error: syntax error in type comment 'Stage -> beam_runner_api_pb2.PTransform'  [syntax]
14:14:31 apache_beam/runners/portability/fn_api_runner/translations.py:1485: note: Suggestion: wrap argument types in parentheses
14:14:31 Found 1 error in 1 file (checked 719 source files)
14:14:31 error: mypy exited with status 2
14:14:32 ERROR: InvocationError for command /home/jenkins/jenkins-slave/workspace/beam_PreCommit_PythonLint_Commit/src/sdks/python/test-suites/tox/py37/build/srcs/sdks/python/target/.tox-py37-mypy/py37-mypy/bin/python setup.py mypy (exited with code 1)
14:14:32 py37-mypy run-test-post: commands[0] | bash /home/jenkins/jenkins-slave/workspace/beam_PreCommit_PythonLint_Commit/src/sdks/python/test-suites/tox/py37/build/srcs/sdks/python/scripts/run_tox_cleanup.sh
14:14:32 ___________________________________ summary ____________________________________
14:14:32 ERROR:   py37-mypy: commands failed

Comment (yifanmai) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py]: Fixed.

**Comment by yifanmai** [comment]
OK, yapf made a number of changes on unchanged lines; let me know if you'd like me to undo them.

**Comment by robertwb** [comment]
Huh, usually yapf doesn't do that. It should be find though.

**Comment by yifanmai** [comment]
Is this ready to merge?

**Pull request merged**